### PR TITLE
Restore allow_unused functionality

### PR DESCRIPTION
**Pull request by ssnl**
Addresses #6535 

I'll do a Variable -> Tensor codemod in a separate PR.

### CODE REVIEW

File: torch/csrc/autograd/python_engine.cpp [hunk 1]
```diff
@@ -154,8 +155,6 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
       if (!grad_fn) {
         output_edges.emplace_back();
       } else {
-        THPUtils_assert(grad_fn,
```
Comment (SsnL) [inline torch/csrc/autograd/python_engine.cpp]: This check was pointless

File: torch/csrc/autograd/python_engine.cpp [hunk 2]
```diff
@@ -92,12 +92,13 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
   unsigned char keep_graph = 0;
   unsigned char create_graph = 0;
   PyObject *inputs = nullptr;
+  unsigned char allow_unreachable = 1;
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: The C default is different than the one in Python. Can we make this consistent?

File: torch/csrc/autograd/python_engine.cpp [hunk 3]
```diff
@@ -172,6 +171,8 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
     THPObjectPtr py_outputs {PyTuple_New(num_inputs)};
     if (!py_outputs) return nullptr;
     for (int i = 0; i < num_inputs; i++) {
+      THPUtils_assert(allow_unreachable || outputs[i].defined(), "One of the differentiated "
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: Can you please change the error message to suggest using the flag if that's what someone wants?

**Review by apaszke — APPROVED** [review]
Two minor things, and it's good to merge.

I actually can't remember why I changed that, sorry to hear that it caused those problems. I know for sure that I wanted to remove `only_inputs`, but I can't think of any reasons why `allow_unused` got cut.

It might have been that I've heard many comments saying that returning `None` is what most people expect, and this is also what other libraries do, but I think it's ok to be strict in this case. This often happens when taking higher order derivatives.

File: torch/csrc/autograd/python_engine.cpp (hunk 2 above)
Comment (SsnL) [reply torch/csrc/autograd/python_engine.cpp]: I set it that way because .backward defaults that to True. But changing this also makes sense. :)

Comment (apaszke) [reply torch/csrc/autograd/python_engine.cpp]: This doesn't matter for `.backward`, right? Is it even an argument?

File: test/test_jit.py [hunk 4]
```diff
@@ -1219,30 +1219,33 @@ def allSum(vs):
...
                             for t in reference_tensors]
 
         outputs_ge = ge(*recording_inputs)
         l1_ge = allSum(outputs_ge)
         grads_ge = torch.autograd.grad(
-            l1_ge, recording_inputs, create_graph=True)
+            l1_ge, recording_inputs, create_graph=True, allow_unused=True)
         l2_ge = (allSum(grads_ge) * l1_ge)
-        grads2_ge = torch.autograd.grad(l2_ge, recording_inputs)
+        grads2_ge = torch.autograd.grad(l2_ge, recording_inputs, allow_unused=True)
```
Comment (apaszke) [inline test/test_jit.py]: Adding `allow_unused` makes those tests much less effective. We need to change the `func` to have a non-zero grad here

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/csrc/autograd/python_engine.cpp (hunk 2 above)
Comment (SsnL) [reply torch/csrc/autograd/python_engine.cpp]: `.backward` also calls `Variable._execution_engine.run_backward`, so it is an arg for the C apt.

File: test/test_jit.py (hunk 4 above)
Comment (SsnL) [reply test/test_jit.py]: The new commit changes it such that we only set `allow_unused` with the `func` that has unused input.

**Review by apaszke — APPROVED** [review]

**Comment by julienroyd** [comment]
Hi @apaszke ,

What is the reason for deprecating the `only_inputs` argument from `grad()`?

I think it was useful to do partial backpropagations while still filling the gradient buffers on the way. Without it, is there another way to halt the backpropagation flow at a given node when calling `backward()` ?

Thanks!

**Pull request merged**